validator_client/src/lib.rs

```diff
@@ -355,13 +354,20 @@ impl<T: EthSpec> ProductionValidatorClient<T> {
         // of making too many changes this close to genesis (<1 week).
         wait_for_genesis(&beacon_nodes, genesis_time, &context).await?;
 
-        let doppelganger_service =
-            (!config.disable_doppelganger_detection).then(|| DoppelgangerService {
-                slot_clock: slot_clock.clone(),
-                validator_store: validator_store.clone(),
-                beacon_nodes: beacon_nodes.clone(),
-                context: context.service_context("doppelganger".into()),
-            });
+        let doppelganger_service = if config.enable_doppelganger_protection {
+            let service = DoppelgangerService::new(
+                context
+                    .service_context(DOPPELGANGER_SERVICE_NAME.into())
+                    .log()
+                    .clone(),
+            );
+
+            validator_store.attach_doppelganger_service(service.clone())?;
```

Comment:
Found an issue where we are still proposing during doppelganger detection. Seems like it's because we are cloning the validator store before creating the doppelganger service, so the block service and attestation service never get a reference to the doppelganger service.

Reply:
Ooft! Great find :grimacing:

I'll knock up a patch for this. I think I'll switch the other services to `Arc<ValidatorStore>` and remove `Clone` for `ValidatorStore`. That type of pattern should help prevent this type of bug from reoccurring.

Reply:
I'm happy to make the patch too if you haven't started, just let me know 🙂

Reply:
The fix over in https://github.com/realbigsean/lighthouse/pull/16/commits/a35bf31e14d08163b7f1d8dacfd5bd62acce2806 looks good, thanks for that! 

I'm wondering why the `Option` for `DoppelgangerService` was removed from `ValidatorStore` and replaced with an "enabled" `bool`? It seems like it would be safer to retain the `Option` since it forces us to check for "enabled" at the type-level. Perhaps I'm missing something :slightly_smiling_face:

Reply:
I removed the `Option` because adding an `Arc` around `ValidatorStore` makes it so we can't attach the `DoppelgangerService` to it after initialization without also adding a `RwLock`. So it was sort of an ergonomics/complexity reduction thing. Let me know if you want me to revert it and add the `RwLock` instead

Reply:
I had a go at it and was able to add the `Option` without a `RwLock` here: https://github.com/paulhauner/lighthouse/commit/15061669595029fafa62c5785f98edffa4710cd8. I also removed `Clone` from `DoppelgangerService` to avoid a similar issue to the one I created with `ValidatorStore`.

I think it's worth having the `Option` since it effectively forces us to check "enabled".

Let me know what you think.

Reply:
Nice! I do like that better!